File: Java/src/deepspace/ShieldBooster.java
/**
 * @brief DEEPSPACE (Java). 
 * Class that represents the shield boosters that space stations can contain
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */

package deepspace;

class ShieldBooster implements CombatElement {
    
    private static final float DEFAULTBOOST = 1.0f;
    
    private String name;
    private float boost;
    private int uses;
    
    /**
     * @brief Constructor with parameters
     * @param theName Name.
     * @param theBoost Boost.
     * @param theUses Use.
     */
    ShieldBooster(String theName, float theBoost, int theUses) {
        this.name = theName;
        this.boost = theBoost;
        this.uses = theUses;
    }
    
    /**
     * @brief Copy constructor
     * @param instance 
     */
    ShieldBooster(ShieldBooster instance) {
        this.name = instance.name;
        this.boost = instance.boost;
        this.uses = instance.uses;
        // this(instance.name, instance.getBoost(), instance.getUses());
    }

    ShieldToUI getUIversion() {
        return new ShieldToUI(this);
    }

    /**
     * @brief decrease the uses of the shield. Override.
     * @pre uses > 0
     * @return boost if uses > 0 or 1.0 in other case
     */
    @Override
    public float useIt() {
        if(uses > 0) {
            uses--;
            return this.boost;
        } else {
            return DEFAULTBOOST;
        }
    } 
    
    /**
     * @brief get method
     * @return attribute boost
     */
    public float getBoost() {
        return this.boost;
    }
    
    /**
     * @brief get method. Override.
     * @return attribute uses
     */
    @Override
    public int getUses() {
        return this.uses;
    }

    public String toString() {
        String data = "[";

        data += "Name:" + this.name + "; ";
        data += "Boost:" + this.boost + "; ";
        data += "Uses:" + this.uses;
        data += "]";

        return data;
    }
}

File: Ruby/ShieldBooster.rb
#encoding:utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | This class specifies the boosters of the shields of each space    | #
# | station.                                                          | #
# +-------------------------------------------------------------------+ #

module Deepspace

    require_relative "ShieldToUI.rb"

    class ShieldBooster

        @@DEFAULTBOOST = 1.0

        # Constructor 
        def initialize (the_name, the_boost, the_uses)
            @name  = the_name
            @boost = the_boost
            @uses  = the_uses
        end
        
        # Copy constructor
        def self.newCopy(instance)
            new(instance.name, instance.boost, instance.uses)
        end

        # Getters

        attr_reader :name, :boost, :uses

        # Decrease the uses of the shield
        def useIt 
            if @uses > 0
                @uses =- 1
                boost
            else
                @@DEFAULTBOOST
            end
        end

        def getUIversion
            return ShieldToUI.new(self)
        end

        # to_s method
        def to_s
            s = "["
            s += "Name: #{name}"
            s += "; Boost: #{boost}"
            s += "; Uses: #{uses}" 
            s += "]"
        end

    end # end of class ShiedlBooster

end # end of Deepspace